### DISCUSSION TITLE
Error calling internal / private functions from Metamask

### DISCUSSION THREAD

**Discussion by tant42** [NomicFoundation/hardhat#3819] [Troubleshooting]
I'm not sure if this is a bug, so wanted to get some understanding first. I'm trying to test my code running an hardhat node and metamask /  web3-onboard / ethers on the browser. 

I have a contract function that calls another contract function conditionally (pseudocode)
```
function postScore(unit score) {
  if (<some address check>) {
    _updateScore(score, msg.sender);
  }
}
```
_updateScore is a private function that does stuff.

When I call from a deploy / test script, all is well. When I call from my browser, I was getting a lot of "function selector not recognized errors".

Added a fallback and things worked again, but the fallback calls were bothering me.  It looked like they were being called by the contract owner (so I assume that was how hardhat was making calls from the contract itself). Since I'm new to digging this deep on these things, I'm assuming these are hardhat's way of doing test calls to estimate gas fees?

After some trial an error, I ended up inlining the code from the private common function into the postScore function and all the errors went away.

I'd like to know:
- Why the errors?
- Is this a bug or a misunderstanding and I need to implement something else in my contracts?

On the "server" / node side:
Hardhat 2.12.4
Hardhat Ethers 2.2.1
Ethers 5.7.2

On the client side:
Ethers 5.7.2

**Comment by fvictorio** [comment]
Hi @tant42. I don't know why you were getting errors in different scenarios, but if I understood correctly you:

- Had a contract function that was working in scripts/tests, but not in the dapp
- Inlining the private function made it work in the dapp.

I see how that looks like a bug but, honestly, it seems very unlikely. If I had to guess, I'd say there was some problem when you tested the dapp (a wrong network, a cache issue on metamask, whatever) and when you inlined the function it coincided with that unrelated problem going away.

That being said, if you can show that this happens consistently (that is, create a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)), then please open an issue about it.
